### PR TITLE
Marc record tensornetwork entropy

### PR DESCRIPTION
**Pull request by SamTov**

### CODE REVIEW

File: CI/unit_tests/observables/test_fisher_trace_calculation.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        ntk = np.array(
+            [
+                [[[1, 2, 3], [4, 5, 6], [7, 8, 9]], np.random.rand(3, 3)],
+                [np.random.rand(3, 3), [[2, 1, 3], [1, 2, 3], [3, 2, 1]]],
+            ]
+        )
+        loss_derivative = np.array([[5, 4, 3], [2, 1, 0]])
+
+        trace = compute_fisher_trace(loss_derivative=loss_derivative, ntk=ntk)
+        assert trace == 638 / 2
```
Comment (SamTov) [inline CI/unit_tests/observables/test_fisher_trace_calculation.py:59]: Where is the 1/2 coming from? Is this just a simple solution to the analytic trace or are you normalising?

File: examples/Computing-Collective-Variables.ipynb
Comment (SamTov) [inline examples/Computing-Collective-Variables.ipynb:1]: Are all of these notebooks cleared of their output cells?

File: znnl/observables/magnitude_entropy.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    Parameters
+    ----------
+    Matrix of which to compute the magnitude entropy.
+
+    Returns
+    -------
+    Magnitude entropy of the matrix."""
+
+    magnitude_dist = compute_magnitude_density(gram_matrix=matrix)
```
Comment (SamTov) [inline znnl/observables/magnitude_entropy.py:47]: Will this be a normalised matrix?

File: znnl/observables/tensornetwork_matrix.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        targets, return_counts=True, return_inverse=True
+    )
+    sortlist = np.argsort(target_indices)
+    # sorting of the ntk
+    sorted_ntk = ntk[sortlist, ...]
+    sorted_ntk = sorted_ntk[:, sortlist, ...]
+    if len(sorted_ntk.shape) == 4:
+        sorted_ntk = np.mean(sorted_ntk, axis=(2, 3))
+
+    """
```
Comment (SamTov) [inline znnl/observables/tensornetwork_matrix.py:56]: This code block should be removed for the merge into main.

File: znnl/observables/tensornetwork_matrix.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+-------
+Module for the computation of the tensornetwork matrix.
+"""
+
+import jax.numpy as jnp
+import numpy as np
+from jax import vmap
+
+
+def compute_tensornetwork_matrix(ntk: np.ndarray, targets: np.ndarray):
```
Comment (SamTov) [inline znnl/observables/tensornetwork_matrix.py:34]: Please add an output type:

```python
def my_func(...) -> np.ndarray:
```

File: znnl/observables/tensornetwork_matrix.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    """
+    Function that calculates the tensornetwork matrix.
+
+    Inputs
+    ------
+    Takes the ntk and the array of targets as an input.
```
Comment (SamTov) [inline znnl/observables/tensornetwork_matrix.py:40]: These should be fully written out with their type and shape

File: znnl/observables/tensornetwork_matrix.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        return jnp.sum(partialntk) / (end1 - start1) / (end2 - start2)
+
+    inner_map = vmap(mapfunc, in_axes=(None, None, 0, 0))
+    outer_map = vmap(inner_map, in_axes=(0, 0, None, None))
+    tensornetwork_matrix = outer_map(startlist, endlist, startlist, endlist)
+
+    return tensornetwork_matrix
+
+
+def _get_start_and_end_value(i, target_counts):
```
Comment (SamTov) [inline znnl/observables/tensornetwork_matrix.py:105]: Perhaps you can put this into the matrix utils module and add a test for it.

File: znnl/training_recording/jax_recording.py [hunk 7]
```diff
@@ -244,7 +276,7 @@ def instantiate_recorder(self, data_set: dict = None, overwrite: bool = False):
         if overwrite:
             self._index_count = 0
 
-        # Check if we need an NTK computation and update the class accordingly
+        # Check if we need an NTK computation, update the class accordingly.
```
Comment (SamTov) [inline znnl/training_recording/jax_recording.py:279]: Why? Or was this the line length issue?

File: znnl/training_recording/jax_recording.py [hunk 8]
```diff
@@ -342,7 +405,7 @@ def visualize_recorder(self):
         -------
 
         """
-        raise NotImplementedError("Not yet available in ZnRND.")
+        raise NotImplementedError("Not yet available in ZnNL.")
```
Comment (SamTov) [inline znnl/training_recording/jax_recording.py:408]: Great catch

File: znnl/training_recording/jax_recording.py [hunk 9]
```diff
@@ -593,6 +701,7 @@ def gather_recording(self, selected_properties: list = None) -> dataclass:
             db_data = self._data_storage.fetch_data(selected_properties)
             # Add db data to the selected data dict.
             for item, data in selected_data.items():
+                print(item)
```
Comment (SamTov) [inline znnl/training_recording/jax_recording.py:704]: Please remove.

File: znnl/training_recording/jax_recording.py [hunk 10]
```diff
@@ -618,3 +727,27 @@ def _export_in_memory_data(self) -> dataclass:
         }
 
         return DataSet(**selected_data)
+
+    def _adjust_ntk_shape(self, parsed_data):
```
Comment (SamTov) [inline znnl/training_recording/jax_recording.py:731]: Do you have tests for this? This can probably also be a matrix util but then it should be moved to the utils module and heavily tested.

**Review by SamTov — COMMENTED** [review]
The code is very well written, nice job. It would be great if you could look at the comments and make some of the changes and we can go from there.
